zfsinstaller/zfsinstaller.cpp

```diff
@@ -23,22 +23,201 @@
   */
 
 #include "zfsinstaller.h"
+#include <ctime>
+#include <string>
 
-// Usage:
-//	zfsinstaller install [inf] [installFolder] (defaults to something like %ProgramFiles%\ZFS)
-//	zfsinstaller uninstall [inf] (could default to something like %ProgramFiles%\ZFS\ZFSin.inf)
-//
-//
+extern "C" {
+	#include <getopt.h>
+	extern char* optarg;
+}
+
+#define MAX_PATH_LEN 1024
+
+//  Usage:
+//    zfsinstaller install [inf] [installFolder] (defaults to something like %ProgramFiles%\ZFS)
+//    zfsinstaller uninstall [inf] (could default to something like %ProgramFiles%\ZFS\ZFSin.inf)
+//    zfsinstaller trace [-f Flags] [-l Levels] [-s SizeOfETLInMB] [-p AbsolutePathOfETL]
+//    zfsinstaller trace -d
+
+const unsigned char ZFSIN_GUID[] = "c20c603c-afd4-467d-bf76-c0a4c10553df";
+const unsigned char LOGGER_SESSION[] = "autosession\\zfsin_trace";
+const std::string ETL_FILE("\\ZFSin.etl");
+
+int session_exists(void) {
+	char   command[256];
+
+	sprintf_s(command, "logman query %s > nul", LOGGER_SESSION);
+
+	return system(command);  // returns 0 if Session exists else non-zero if Session does not exist
+}
+
+int zfs_log_session_delete(void) {
+	char command[MAX_PATH_LEN];
+
+	int ret = session_exists();
+
+	if (ret == 0) {  // Session exists
+		sprintf_s(command, "cmd.exe /c \"logman delete %s > nul\"", LOGGER_SESSION);
+		ret = system(command);
+		return ret;
+	}
+	else {
+		return 0; // Session does not exist ; We will pass success
+	}
+}
+
+int validate_args(long long int flags, int levels, int size_in_mb, const char* etl_file) {
+	if (flags < 0 || flags > 0xffffffff) {
+		fprintf(stderr, "Valid input for flags should be in interval [0, 4294967295]\n");
+		return 1;
+	}
+
+	if (levels < 0 || levels > 255) {
+		fprintf(stderr, "Valid input for levels should be in interval [0, 255]\n");
+		return 2;
+	}
+
+	if (!strcmp(etl_file, "") || !strstr(etl_file, ".etl")) {
+		if (etl_file) {
+			fprintf(stderr, "Etl file path/name %s is incorrect\n", etl_file);
+		}
+		else {
+			fprintf(stderr, "Etl file path/name is incorrect\n");
+		}
+		return 3;
+	}
+
+	if (size_in_mb <= 0) {
+		fprintf(stderr, "Size of etl should be greater than 0\n");
+		return 4;
+	}
+	return 0;
+}
+
+int move_file(const char *etl_file) {
+	char move_etl[MAX_PATH_LEN];
+	time_t rawtime;
+	struct tm timeinfo;
+	char buffer[25];
+
+	time(&rawtime);
+	localtime_s(&timeinfo, &rawtime);
+	strftime(buffer, sizeof(buffer), "_%Y%m%d%H%M%S", &timeinfo);
+
+	strcpy_s(move_etl, etl_file);
+	char* etl = strstr(move_etl, ".etl");
+	if (etl)
+		etl[0] = 0;
+
+	strcat_s(move_etl, buffer);
+	strcat_s(move_etl, ".etl");

```

Comment:
idem

Reply:
buffer would have string "_YYYYMMDDHHMMSS";
line 112 will append buffer to $PATH\\$ETL_NAME and at last ".etl" is appended to it